File: src/engine-config.ts
import camelCase from "lodash.camelcase";
import { EngineConfig, Section } from "./definition";
import { parseSection } from "./section-parser";

function newEngineConfig(): EngineConfig {
	return {
		sound: 1,
		battle: 1,
		drama: true,
		autoSave: true,
		stretch: false,
		campaign: 0,
		host: "localhost",
		textSpeed: 1000,
		formLeft: 0,
		formTop: 0,
		formMaximum: 0,
	};
}

function parse(section: Section): EngineConfig {
	if (section.name !== "CONFIG") {
		throw new Error("not a engine config");
	}

	const conf = newEngineConfig();
	section.lines.forEach(line => {
		const [k, v] = line.split('=');
		switch (k) {
			case "Sound":
			case "Battle":
			case "Campaign":
			case "TextSpeed":
			case "FormLeft":
			case "FormTop":
			case "FormMaximum":
				conf[camelCase(k)] = parseInt(v, 0);
				return;
			case "Drama":
			case "AutoSave":
			case "Stretch":
				conf[camelCase(k)] = v !== "0";
				return;
			case "Host":
				conf[camelCase(k)] = v;
				return;
			default:
				conf[k] = v;
				return;
		}
	})

	return conf;
}

function load(input: string) {
	const result = parseSection(input);
	return parse(result[0]);
}

export default {
	load,
	parse,
	newEngineConfig,
};

File: engine_config.go
package main

import (
	"log"
	"strconv"
	"strings"
)

type EngineConfig struct {
	Sound     bool  `json:"sound"`
	Battle    bool  `json:"battle"`
	Drama     bool  `json:"drama"`
	AutoSave  bool  `json:"autoSave"`
	Stretch   bool  `json:"stretch"`
	Campaign  int64 `json:"campaign"`
	TextSpeed int64 `json:"textSpeed"`
}

func (conf *EngineConfig) convert(section *Section) *EngineConfig {
	if section.name != "CONFIG" {
		log.Fatalln("Not a engine config")
	}

	for _, line := range section.lines {
		pairs := strings.Split(line, "=")
		k, v := pairs[0], pairs[1]
		switch k {
		case "Sound":
			conf.Sound = v == "1"
		case "Battle":
			conf.Battle = v == "1"
		case "Drama":
			conf.Drama = v == "1"
		case "AutoSave":
			conf.AutoSave = v == "1"
		case "Stretch":
			conf.Stretch = v == "1"
		case "Campaign":
			conf.Campaign, _ = strconv.ParseInt(v, 10, 64)
		case "TextSpeed":
			conf.TextSpeed, _ = strconv.ParseInt(v, 10, 64)
		}
	}

	return conf
}
